### DISCUSSION TITLE
Finetuning models for 7-segment fonts

### DISCUSSION THREAD

**Discussion by dchaplinsky** [mindee/doctr#1467] [General]
One task we are currently trying to solve is to recognise OSD texts on video, made with a 7 segment font (mostly digits). Those overlays are usually well made and contrast, so it is not a text in the wild in a wide sense.

What I've found is that almost all the detectors are failing to find the correct bounding boxes for such texts (even when using low binarization threshold and the best models, like `db_resnet50`). If I provide the correct bbox myself (thanks again for the callbacks, @felixdittrich92!) the parseq recognizer can pick that (but also produces unstable result). Other recognition architectures like `crnn_vgg16` has even worse performance.

To illustrate my point I made some synthetic examples:
<img width="1242" alt="Screenshot 2024-02-13 at 13 28 57" src="https://github.com/mindee/doctr/assets/131186/1c3e0654-d91d-4ed8-bc64-44533e43a750">
<img width="1242" alt="Screenshot 2024-02-13 at 13 35 53" src="https://github.com/mindee/doctr/assets/131186/f79a86e1-7d95-4d48-aa62-96eded561572">

So even with the best models on a very clean document it produces mediocre result.
I've also tried mmocr, but both me and my colleague failed to make it build and run, in general project seems to be in a very neglected state, which is shame.

So my rough plan is following:

- Generate some synthetic dataset which utilizes 7 segment fonts and various backgrounds (document-white, wild, etc).
- Train better detector
- Train better recognizer models
- Share them (if my boss is ok with that).

I can also try adding some existing datasets like TotalText / ICDAR into the training data.

Now the big questions is: how do I train? I do not want to loose the existing models, as they work perfectly fine for the majority of other videos we have. I also cannot train those from ground zero with your data, as your training datasets are not public (which is understandable). 
Can I finetune existing architectures? I saw that training scripts can pick the existing models, but I do not have an experience with finetuning of CV models. Should I try low LR and some LR schedule with warmup and decay, to not make model forget the things it learned from the past?

**Comment by felixdittrich92** [comment]
Hey @dchaplinsky :wave: 

It looks like that the images you try to process are relative small is this correct ?

About detection:

- freezing the backbone makes no sense in my mind because this would only work for fine tuning on samples which are more close to our training set 
- from our experiments the current default scheduler PolyLR works best
- about the LR i would give it a try with --pretrained and a lr of 0.0001 (epochs depends on the number of samples - less data => more epochs)
- about wd should be set to 0.1 without trouble

About recognition:

- I have also fine tuned `parseq` completely on synthetic data created with https://github.com/clovaai/synthtiger
- Everything as is (default values) only with the new data and vocab

**Reply by dchaplinsky** [reply]
Thanks, will give it a try!

**Comment by dchaplinsky** [comment]
@felixdittrich92 thanks again for the helpful guide.
I've started with re-running the evaluation using `db_resnet50` and `CORD`, `FUNSD` dataset to get the baseline, so I can control the degradation of the model after finetune on my data.

However, results differs a lot from the one reported [here](https://mindee.github.io/doctr/latest/using_doctr/using_models.html#text-detection).
<img width="645" alt="Screenshot 2024-02-21 at 15 17 15" src="https://github.com/mindee/doctr/assets/131186/9cd7bce6-a309-45e2-946c-2ec725b23029">

For example:
CORD recall/precision: `92.61 | 86.37` vs `92.96 | 91.23` on a model page.
FUNSD recall/precision: `83.55 | 86.67` vs `79.17 | 86.31` on a model page.

I'm using pytorch version (I slightly modified it to allow faster eval on apple silicon, probably wrap it into the PR) like this:
`time python evaluate_pytorch.py db_resnet50 --dataset CORD`
`time python evaluate_pytorch.py db_resnet50`

No other modifications were made. 
Any ideas, why it differs?

**Comment by felixdittrich92** [comment]
Hi @dchaplinsky 👋 

The actual Benchmarks https://github.com/mindee/doctr/pull/1321

**Reply by dchaplinsky** [reply]
Thanks! Can't wait for the release!

**Reply by dchaplinsky** [reply]
What vocabulary are you using for recognition eval? default (french)?

**Reply by felixdittrich92** [reply]
Correct :)

**Reply by dchaplinsky** [reply]
Hmmmm, what I am doing wrong then?
```
python evaluate_pytorch.py crnn_vgg16_bn
Validation loss: 41.0507 (Exact: 0.00% | Partial: 0.00%)
```

```
python evaluate_pytorch.py crnn_vgg16_bn --only_regular
Validation loss: 38.2464 (Exact: 0.00% | Partial: 0.00%)
```

**Reply by felixdittrich92** [reply]
On main branch ? Check that the model is already in the `.cache` folder and pretrained is True .. i remember there was a issue with the model loading if the file isn't already downloaded :)

**Reply by dchaplinsky** [reply]
Yes, on main branch (pulled today).
```
% ls -lh ~/.cache/doctr/models/crnn_vgg16_bn-9762b0b0.pt
-rw-r--r--@ 1 dchaplinsky  staff    60M Dec  2 01:32 /Users/dchaplinsky/.cache/doctr/models/crnn_vgg16_bn-9762b0b0.p
```

And this is the gibberish it returns:
```
{'preds': [(';<H(S`CyS`2SYaHLl<+HP', 0.012497764080762863)], 'loss': tensor(44.0452)}
{'preds': [('Wq=qE3xq|n<=5<KvaHvH', 0.012593762017786503)], 'loss': tensor(26.0040)}
{'preds': [('SqKvaHq<+Hq', 0.012921293266117573)], 'loss': tensor(143.0937)}
{'preds': [('aqvSwy`y1Sp1H฿`WS(aP`xHVPH`', 0.01302513387054205)], 'loss': tensor(21.0756)}
{'preds': [(';q3VS<฿<SKSyYnaHxH8H', 0.012707684189081192)], 'loss': tensor(31.7654)}
```

**Reply by dchaplinsky** [reply]
Last commit is:
```
commit dd1fbbe69903321f30a86aec264b491038d57a30 (HEAD -> main, origin/main, origin/HEAD)
Author: Felix Dittrich <felixdittrich92@gmail.com>
Date:   Thu Feb 15 08:55:49 2024 +0100

    [Fix] Catch Divide by zero explicit (#1471)
```

I also deleted and downloaded model again, that didn't help. Model is loaded as `pretrained=True`, dataset is using `recognition_task=True`

**Reply by felixdittrich92** [reply]
Ok that's weird.. Same result if you use --resume=<PATH_TO_CACHE/crnn_vgg...pt> ?

I run it with:
`USE_TORCH=1 python3 /home/felix/Desktop/doctr/references/recognition/evaluate_pytorch.py crnn_vgg16_bn --dataset=FUNSD`

**Reply by dchaplinsky** [reply]
RuntimeError: Error(s) in loading state_dict for CRNN:
        size mismatch for linear.weight: copying a param with shape torch.Size([124, 256]) from checkpoint, the shape in current model is torch.Size([101, 256]).
        size mismatch for linear.bias: copying a param with shape torch.Size([124]) from checkpoint, the shape in current model is torch.Size([101]).
        
Nope :( 
I'll try running it on proper GPU.

**Reply by felixdittrich92** [reply]
--vocab=legacy_french

This model is an old one trained on the legacy_french vocab instead of the french vocab sry i missed to mention this

**Reply by dchaplinsky** [reply]
`master` and `parseq` also gave me some garbage.

legacy_french helped tho:
`{'preds': [('PROVIDENCE', 0.9800896644592285)], 'loss': tensor(0.0020)}`.

Should I use legacy_french to all of them?

**Reply by felixdittrich92** [reply]
No only crnn_vgg16_bn is still trained on the old vocab all other default to french

**Reply by felixdittrich92** [reply]
Check the cache again for the other models :)

**Reply by dchaplinsky** [reply]
That is very strange. parseq on CPU (MBA M2, no MPS) gives me total gibberish (both french and legacy_french), same code and models (I shasumed them) on NVidia T4 works flawlessly on default params, the eval results are almost identical to yours.
Not a big deal, just puzzling.

**Reply by dchaplinsky** [reply]
Ok, ignore me, I changed the default value for the vocab in the eval script to english and totally forgot it. Now everything works ok.

**Comment by dchaplinsky** [comment]
Well, I'm definitely getting somewhere.

At least can confirm that it trains on synthetic data I generated and have a basic understanding on the scale of dataset.

<img width="1321" alt="Screenshot 2024-03-15 at 19 05 22" src="https://github.com/mindee/doctr/assets/131186/ea35930f-f6ed-40e9-b23c-078ecc7c1e2c">

Next stop: finetune of your checkpoints.
